op-node/rollup/derive/channel_test.go

```diff
@@ -99,3 +106,93 @@ func TestFrameValidity(t *testing.T) {
 		t.Run(tc.name, tc.Run)
 	}
 }
+
+func TestBatchReader(t *testing.T) {
+	// Get batch data
+	rng := rand.New(rand.NewSource(0x543331))
+	singularBatch := RandomSingularBatch(rng, 20, big.NewInt(333))
+	batchDataInput := NewBatchData(singularBatch)
+
+	encodedBatch := &bytes.Buffer{}
+	buf := &bytes.Buffer{}
+	// Get the encoded data of the batch data
+	err := batchDataInput.encodeTyped(buf)
+	require.NoError(t, err)
+	err = rlp.Encode(encodedBatch, buf.Bytes())
```

Comment:
I still don't understand why this is needed. Looking at `span_channel_out` it does `rlp.Encode()` on `spanBatch` object not the encoded spanBatch bytes.  https://github.com/ethereum-optimism/optimism/blob/develop/op-node/rollup/derive/span_channel_out.go#L143

So in `span_channel_out` it it's
```
spanBatch => rlp.Encode() => zlib
```

but here we are doing
```
spanBatch => rlp.Encode() => rlp.Encode(buf) => zlib
```

Reply:
I think you could actually remove the `buf` and directly call `batchDataInput.EncodeRLP(encodedBatch)`. The confusion might come from this. Span (or singular) batches internally don't already do the rlp compression. The `InnerBatchData` is implemented e.g. by the `RawSpanBatch`. If you check its `encode` function, it doesn't do rlp (besides the rlp encodings that are part of the internal span batch spec).

Also check out the test `TestBatchRoundTripRLP` in `batch_test.go`, it looks quite similar to this one. Maybe we can deduplicate. In any case, would be great to also add Brotli tests to `TestBatchRoundTripRLP`. It also uses `EncodeRLP`.

Reply:
Note also that you're not using a `SpanChannelOut` in this test, but basically duplicate some internal logic of it, but skip the frame-chopping part.

Reply:
I see interesting, yeah just tried with `batchDataInput.EncodeRLP(encodedBatch)` which works too.

It looks like `TestBatchRoundTripRLP` mainly tests for RLP encoding/decoding, I think doing a full round trip in `TestBatchReader` makes more sense imo.

I can add another test that uses `SpanChannelOut` and frames, but perhaps in a follow up PR along with the e2e test change? So we can merge this PR first

Reply:
@cody-wang-cb I think a full E2E test in a follow-up is sufficient, thanks!